Adding OBELICS DataLoader

TJ-Solergibert opened the pull request:
Hi,

In this PR I present a first draft of the Multimodal DataLoader. First I will describe how the batches are created and then I will explain the padding problem. 

<img width="1106" alt="image" src="https://github.com/user-attachments/assets/8dd684e6-53e8-4248-b1a7-edbae846d5d4">

Let's begin checking the [OBELICS](https://huggingface.co/datasets/HuggingFaceM4/OBELICS) dataset. For every sample on the dataset we have 4 keys, but we are just interested in 2 of them:

- `images`: A list either with URLs of images OR `None`s to specify the position of the text.
- `texts`: A list either with text strings OR `None`s to specify the position of the images.
It's important to highlight that `len(images)==len(texts)` and that for each index, **one element and only one** is not `None`. 

The [`format_obelics`](https://github.com/TJ-Solergibert/torchtitan/blob/07a7a12af64075225874b18c99ab158776c2d254/torchtitan/datasets/multimodal/utils.py#L438) function will transform each sample to a format that can be later fed into the transform block that will prepare the samples to the target type. Each formatted sample will be a dictionary containing 2 keys:
- `images`: `List` of PIL Images with the loaded images.
- `text`: `str` with the text of the sample ready to be tokenized, including the image tokens.

<img width="1302" alt="image" src="https://github.com/user-attachments/assets/05a20daa-efb6-4bc7-be21-722a93e19f48">

Once formatted, we will process each sample with the transform block. This transform block is composed of [`CLIPPreprocess`](https://github.com/TJ-Solergibert/torchtitan/blob/07a7a12af64075225874b18c99ab158776c2d254/torchtitan/datasets/multimodal/clip.py#L25), `TikTokenizer` & [`VisionCrossAttentionMask`](https://github.com/TJ-Solergibert/torchtitan/blob/multimodal_dl/torchtitan/datasets/multimodal/vision_attention_mask.py) modules. 

## `CLIPPreprocess`
<img width="1293" alt="image" src="https://github.com/user-attachments/assets/1e19c5cf-f17d-4954-8bbc-6de2b5bb72be">

This module will prepare the List of images to be fed into the CLIP model. The most relevant steps is resizing the image without distortion, dividing the image into tiles and padding if necessary. Highlight the fact that it will still produce a List of tensors and NOT a tensor as every image can have a different number of tiles. This will be addressed in the collator where we will pad the image tiles to the largest in the batch. Also, we keep the maximum number of tiles to 4 and the tile size to 448 for pretraining [[1]](https://github.com/pytorch/torchtune/blob/0cc1b1f6a2a9c54ca640c4eb0a4d0b94ba94bb04/torchtune/models/llama3_2_vision/_model_builders.py#L92), [[2]](https://huggingface.co/meta-llama/Llama-3.2-11B-Vision/blob/3f2e93603aaa5dd142f27d34b06dfa2b6e97b8be/preprocessor_config.json#L22). 

## `TikTokenizer`
I've included a new method in the tokenizer to encode the multimodal text. In short, it just encodes the text adding the special `image_id` token and returns both the `input_ids` & `labels` masking the `bos`, `eos` & `image_id` tokens.

## `VisionCrossAttentionMask `
<img width="1298" alt="image" src="https://github.com/user-attachments/assets/a163a69f-1225-44bd-add1-e7e8e06e81fd">

This module will create the attention mask for the Fused layers. In short, for each **TILE** we will have 1025 `image_tokens` and this mask will specify for each `text_token` to which `image_tokens` should attend to. We are returning again a List of tensors as the quantity of `image_tokens` will depend on the number of tiles. Again, we will solve this in the collator.

# Padding & the collator
As we've previously seen, both the outputs of the `CLIPPreprocess` & `VisionCrossAttentionMask` are list of tensors because of the different number of tiles. Within the same sample we should pad both artifacts to the maximum number of tiles, but the issue arises when we run `batch_size > 1` as we will also need to pad the `input_ids` (& `labels`) which is relatively cheap BUT also the Number of images, as the input to the CLIP model will be a tensor of shape [Batch size, Number of images, Number of tiles, Channels, Tile size, Tile size]. Padding to the maximum number of tiles is bad, but in the worst case scenario you end up increasing the tensor x4 (from 1 tile to maximum number of tiles = 4). But for the number of images it can get really really big, as there are samples with +30 images. 

To check this phenomenon I've included [`scripts/check_padding_mm.py`](https://github.com/TJ-Solergibert/torchtitan/blob/multimodal_dl/scripts/check_padding_mm.py) which computes the % of padding in a sample. Feel free to give it a try but it's very easy to get samples where the majority of the input is padding.

```
python3 scripts/check_padding_mm.py
Unpadded tokens: 8717, Total tokens in batch: 21728
Padded text tokens: 13011, 59.88%
########################################
Unpadded images: 25, Total images in batch: 64
Padded images: 39, 60.94% (Each image with shape [4, 3, 448, 448])
########################################
Unpadded number of tiles: 61, Total number of tiles: 256
Padded tiles: 195, 68.72% (Each with shape [3, 448, 448])
########################################
Unpadded cross attention mask elements: 545030425, Total cross attention mask elements: 5701427200
Padded cross attention mask elements: 5156396775, 90.44%
```

That's why I proposed continue working on a DataLoader & Dataset than can pack multiple samples up to a given `input_ids` length OR number of images in a batch. Packing the `input_ids` is fairly easy while packing the cross attention masks will require a bit more effort. Let me know if you would be interested on supporting that feature or you just want to include in the repo an example of the multimodal pipeline despite the padding issue described. I also plan including some unit test, to check the generated samples & recovering from failures abilities. 

Other comments:
- [torchtitan/datasets/mm_datasets.py](https://github.com/TJ-Solergibert/torchtitan/blob/multimodal_dl/torchtitan/datasets/mm_datasets.py): I hardcoded the image token & also the Llama3VisionTransform init. Let me know whether I should leave this (default) values or let the user modify them.
- [torchtitan/datasets/multimodal/collator.py](https://github.com/TJ-Solergibert/torchtitan/blob/multimodal_dl/torchtitan/datasets/multimodal/collator.py): I hardcoded the ignore index to mask the special tokens & added some debug lists for the  [`scripts/check_padding_mm.py`](https://github.com/TJ-Solergibert/torchtitan/blob/multimodal_dl/scripts/check_padding_mm.py) script.
- [torchtitan/datasets/multimodal/llama3_transform.py](https://github.com/TJ-Solergibert/torchtitan/blob/multimodal_dl/torchtitan/datasets/multimodal/llama3_transform.py): More hardcoded values & what should we do (Include and mask) with eos & bos tokens.
- [torchtitan/datasets/multimodal/utils.py](https://github.com/TJ-Solergibert/torchtitan/blob/multimodal_dl/torchtitan/datasets/multimodal/utils.py): Most of the code directly copied from `torchtune` cleaning the unnecessary parts like the code for the inference case. Also in the [`format_obelics`](https://github.com/TJ-Solergibert/torchtitan/blob/07a7a12af64075225874b18c99ab158776c2d254/torchtitan/datasets/multimodal/utils.py#L438) function we could drop the last images in the case the sample end with images and not text as no token will attend to them and we dont compute the loss with the image tokens (So they are useless)
- [torchtitan/datasets/tokenizer/tiktoken.py](https://github.com/TJ-Solergibert/torchtitan/blob/multimodal_dl/torchtitan/datasets/tokenizer/tiktoken.py): Hardcoded image token + encode_multimodal method. Is it fine to include this method here or should we move it somewhere else? Also we could standardise the nomenclature for the `input_ids`/`tokens` across the repo.

Toni

fduwjj commented on torchtitan/datasets/multimodal/clip.py:
```diff
@@ -0,0 +1,195 @@
+
+import torch
+
+import torchvision
```
This is just a quick scan for this PR. I need more time to look into this PR. But we should add torchvision as the dependent of Torchtitan, if haven't.

TJ-Solergibert replied:
Yes! Don't worry, it will take time 😅

fduwjj commented on scripts/check_padding_mm.py:
```diff
@@ -0,0 +1,70 @@
+BATCH_NUMBER = 4
+
+
+def main():
```
maybe we can make this as a unit test? WDYT?

TJ-Solergibert replied:
I would add as a unit test some checks of shapes & types on the DP axis rather than this script that just checks the amount of padding in each batch

fduwjj commented on .pre-commit-config.yaml:
Is this file expected?

fduwjj commented on torchtitan/datasets/tokenizer/tiktoken.py:
```diff
@@ -190,3 +195,30 @@ def _split_whitespaces_or_nonwhitespaces(
+            text, bos=True, eos=True, allowed_special=set(["<|image|>"])
+        )
+        input_ids = tokens[:-1]
+        labels = tokens[1:]
```
what if the first token is null? And for `HuggingFaceM4/OBELICS` specifically, "text" is a list of string or null. We don't need special treatment of it?

TJ-Solergibert replied:
We take care of the null/None values in the [`format_obelics`](https://github.com/TJ-Solergibert/torchtitan/blob/07a7a12af64075225874b18c99ab158776c2d254/torchtitan/datasets/multimodal/utils.py#L438) function. 

This function produces:
- `images`: List of decoded images in the sample
- `text`: str with the text of the sample ready to be tokenized **including the image tokens** (the null values you are referring)

So the `text` is ready to be tokenized!

fduwjj commented on torchtitan/datasets/hf_datasets.py:
```diff
@@ -156,8 +156,14 @@ class DPAwareDataLoader(StatefulDataLoader, Stateful):
+        batch_size: int,
+        collator_fn: Callable,
+    ):
+        super().__init__(dataset=hf_ds, batch_size=batch_size, collate_fn=collator_fn)
```
where is this `collate_fn` being used or called?

TJ-Solergibert replied:
Now we also have #1021 which is also adding the collator. For the HF datasets solution already in torchtitan we dont need it as the samples produced by the Dataset are ready to go. For the multimodal one we need it to pad to the longest/biggest samples, otherwise, we can directly pad to the longest supported shapes possible directly in the dataset (I prefer the collator solution)

fduwjj commented on torchtitan/datasets/multimodal/collator.py:
```diff
@@ -0,0 +1,289 @@
+            ...         "encoder_mask": [torch.ones(2, 5 * 4)],
+            ...     },
+            ... ]
+            >>> model_inputs = padded_collate_tiled_images_and_mask(batch=batch)
```
Is this the way this collator is used?

fduwjj commented on torchtitan/datasets/multimodal/collator.py:
```diff
@@ -0,0 +1,289 @@
+
+# NOTE Inspired from torchtune.data._collate.py
+@dataclass
+class MultiModalCollator:
```
IIUC, this generates the all the data needed before sending into encoder right? Also for the MM model, we need to feed text tokens into decoder as well? shall we just reuse the existing dataloader for llama3?

TJ-Solergibert replied:
> shall we just reuse the existing dataloader for llama3?
No! Now, with the Dataset & DataLoader in `mm_dataset.py` & `mm_dataloader.py` + the collator we prepare all the inputs with a single DataLoader!

The collator returns the prepared batches with:
```python
batch_dict = {
  "tokens": collated_text["tokens"],
  "labels": collated_text["labels"],
  "encoder_input": {
      "images": collated_images,
      "aspect_ratio": collated_aspect_ratios,
  },
  "encoder_mask": concat_masks,
}
```

fduwjj reviewed (commented):
Sorry for super late review. I finally finish the first round of review and let me know what do you think about my questions and comments. I will revisit this PR in the coming week.

TJ-Solergibert commented:
Sorry, I'll try to address all this comments in the next 10 days!

fduwjj commented:
@TJ-Solergibert I am wondering if we can split this PR and get it merged in pieces?

tianyu-l commented:
Hey @TJ-Solergibert , are you still interested in continuing to work on this PR?

TJ-Solergibert commented:
Updated the PR to the main branch, incorporating new features from torchtitan like the `TrainSpec`. I'll address the comments in the following 2 days!

TJ-Solergibert commented on torchtitan/datasets/tokenizer/tiktoken.py:
```diff
@@ -67,6 +70,7 @@ def __init__(self, model_path: str):
         self.special_tokens = {
             token: num_base_tokens + i for i, token in enumerate(special_tokens)
         }
+        self.special_tokens["<|image|>"] = 128256  # TODO(tj.solergibert) Hardcoded!
```
Now we have `IMAGE_TOKEN_ID` in `datasets/tokenizer/tiktoken.py` and we use it in `experiments/multimodal/llama3_transform.py` for `VisionCrossAttentionMask`

TJ-Solergibert commented on torchtitan/datasets/multimodal/clip.py:
```diff
@@ -0,0 +1,195 @@
+            Mapping[str, Any]: The sample with an updated "image" filed and added
+                "aspect_ratio" field.
+        """
+        image = sample["image"]
```
The docs in `_process_obelics_sample` state clearly the structure that a sample_processor generates in order to work with `Llama3VisionTransform`. I guess that we can mark this comment as resolved now that we have the sample_processor & text_processor + datasets.md in docs, right?

TJ-Solergibert commented on torchtitan/datasets/multimodal/clip.py:
```diff
@@ -0,0 +1,195 @@
+    best_resolution = (448, 1344) # canvas that allows maximum upscaling, with minimum padding, up to 16 tiles
+    image is resized without distortion (300,800) -> (448, 1194) #448 is the limiting side for the resize
+    image is padded (448, 1194) -> (448, 1344)
+    Image is tiled 2x5, for a final output shape of (10, 3, 224, 224)
```
Yup, you are right, it's still wrong in [torchtune](https://github.com/pytorch/torchtune/blob/c5c160b421b47d5257ad270015e13b7c38543127/torchtune/models/clip/_transform.py#L49)

TJ-Solergibert commented on torchtitan/datasets/multimodal/clip.py:
```diff
@@ -0,0 +1,195 @@
+        max_num_tiles (Optional[int]): Only used if possible_resolutions is NOT given.
+            Maximum number of tiles to break an image into.
+            This will be used to generate possible_resolutions,
+            e.g. [(224, 224), (224, 448), (448, 224)] if max_num_tiles = 2 and tile_size = 224.
```
Tiles have length and height, 2D. (448, 448) is 4 tiles, while the 3 examples mentioned in the example have only 2 ((224, 224) single tile, (224, 448) one tile next to another & (448, 224) one tile up another down). 

For `max_num_tiles` = 4 we have: [(224, 896), (448, 448), (224, 224), (896, 224), (224, 672), (672, 224), (224, 448), (448, 224)]. 

We can put this example in the docstring but it's larger.

TJ-Solergibert commented on torchtitan/datasets/multimodal/clip.py:
```diff
@@ -0,0 +1,195 @@
+                "aspect_ratio" field.
+        """
+        image = sample["image"]
+        assert isinstance(image, Image.Image), "Input image must be a PIL image."
```
In torchtune since this [PR](https://github.com/pytorch/torchtune/pull/2366) now they return tensors instead of PIL Images. They leverage `torchvision.io.decode_image(...)`. I will migrate to torchvision too.

TJ-Solergibert commented on torchtitan/datasets/multimodal/utils.py:
```diff
@@ -0,0 +1,458 @@
+    ]
+    return {
+        "images": [load_image(image) for image in sample_images],
+        "text": "".join(map(str, sample_text)),
```
Maybe you are missing `VisionCrossAttentionMask`?

TJ-Solergibert commented on torchtitan/datasets/multimodal/utils.py:
```diff
@@ -0,0 +1,458 @@
+    that contain up to max_num_tiles. Useful for when dividing an image into tiles.
+
+    For example, if we want at most 2 tiles per image, then we can support the
+    following resolutions: (1x1, 1x2, 2x1) * tile_size
```
Tiles have 2 dimensions!

![Tiles](https://github.com/user-attachments/assets/1b67d7ed-6f10-435f-a466-49a5cad4de0c)

Now it's clear?

TJ-Solergibert commented on torchtitan/datasets/multimodal/utils.py:
```diff
@@ -0,0 +1,458 @@
+    then the maximum upscaling without distortion is min(2, 1.5) = 1.5.
+
+    If there are multiple canvases that satisfy the conditions,
+    we pick the one with the lowest area to minimize padding.
```
You want to resize without distortion. From the example in the docstring, given the `possible_resolutions` you end up having this 2 possibilities, one with 2 tiles and another one with 3. We pick the one with just 2 tiles to minimise padding.

![Padding](https://github.com/user-attachments/assets/3686cf13-823a-4a84-b8bf-b38f646b27a4)

TJ-Solergibert commented on torchtitan/datasets/mm_datasets.py:
```diff
@@ -0,0 +1,144 @@
+                ].count(self.image_token)
+                self._sample_idx += 1
+                # Transform sample
+                processed_sample = self.transform(processed_sample)
```
Changed to `self.format(processed_sample)` & `Llama3VisionFormatter`

TJ-Solergibert commented on torchtitan/datasets/multimodal/vision_attention_mask.py:
```diff
@@ -0,0 +1,171 @@
+            >>> transform = VisionCrossAttentionMask(tile_size=400, patch_size=40, image_token_id=1)
+            >>> intervals = transform._get_image_attention_intervals(tokens)
+            >>> print(intervals)
+            [[0, 7], [1, 7], [7, 12]]
```
Text tokens only attend to the previous image, multiple images if they are consecutive. 

From the example: "<img1><img2>These are two dogs. <img3>This is a cat."

`These are two dogs.` will attend to <img1> & <img2> and "This is a cat." only to <img3>

pbontrager reviewed (commented):
Thank you for the hard work here. This looks good but includes more Llama3.2 code than I think we need to enable MM in torchtitan. Since most modern VLM's use Early Fusion architectures instead of Deep Fusion like 3.2, we should choose to just support Early Fusion models for now. I left some comments on what could be removed or moved. After that it looks good to go.

TJ-Solergibert commented on torchtitan/experiments/multimodal/utils.py:
```diff
@@ -0,0 +1,458 @@
+    return image
+
+
+def format_obelics(sample: Dict, image_token: str = "<|image|>") -> Dict:
```
Deleted, missed to push the change sorry!

TJ-Solergibert commented on torchtitan/experiments/multimodal/clip.py:
```diff
@@ -0,0 +1,195 @@
+from torchvision.transforms.v2 import functional as F
+
+# NOTE Inspired from torchtune.models.clip._transform.py
+class CLIPPreprocess:
```
I rename it because of [this](https://github.com/pytorch/torchtitan/pull/663#discussion_r1901376612) previous comment

TJ-Solergibert commented on torchtitan/experiments/multimodal/mm_dataset.py:
```diff
@@ -0,0 +1,210 @@
+        self._sample_processor = sample_processor
+        self.image_token = "<|image|>"  # TODO(tj.solergibert) Hardcoded!
+
+        self.transform = Llama3VisionTransform(
```
Left multiples. We have to decide which variables we want to expose through JobConfig

TJ-Solergibert commented on torchtitan/experiments/multimodal/llama3_transform.py:
```diff
@@ -0,0 +1,101 @@
+from vision_attention_mask import VisionCrossAttentionMask
+
+# NOTE Inspired from torchtune.models.llama3_2_vision._transform.py
+class Llama3VisionTransform:
```
Renamed to `VisionFormatter`

TJ-Solergibert replied:
[Removed](https://github.com/pytorch/torchtitan/pull/663#discussion_r2019527153)

pbontrager commented on torchtitan/experiments/multimodal/llama3_transform.py:
```diff
@@ -0,0 +1,269 @@
+        return sample
+
+
+class VisionFormatter:
```
Can we go back to the original name for this? I think we should either match the name in torchtune Llama3VisionTransform or just call it a tokenizer. @tianyu-l do you have a preference for this name?

tianyu-l replied:
Can I ask why we combine tokenizing and image transform into one class? It looks to me not having efficiency benefit, but mixing independent functionalities into one class, (and as a side-effect making naming hard). Can we split tokenizing out? I even think this class may not be necessary, can we merge it into `MultiModalDataset`?

pbontrager reviewed (commented):
This looks good, thank you for the quick turnaround. I left one additional comment, but I'd be happy to land it now and then we can iterate on it further in follwup PRs. If you're ready to land it, you can just remove your personal todo comments and remove [WIP] and I'll approve it.

tianyu-l commented on torchtitan/datasets/tokenizer/tiktoken.py:
Can we create a dedicated .py file in the experiments folder? Let's keep the MM changes in experiments for now.

tianyu-l commented on torchtitan/components/dataloader.py:
```diff
@@ -73,7 +73,7 @@ def load_state_dict(self, state_dict: dict[str, Any]) -> None:
 
         if self._rank_id not in state_dict:
-            logger.warning(
+            logger.warning(  # NOTE(tj.solergibert) Missing import?
```
this has been addressed, please remove comment

tianyu-l commented on torchtitan/experiments/multimodal/check_padding_mm.py:
```diff
@@ -0,0 +1,69 @@
+
+from mm_dataset import build_mm_dataloader
+
+PATH_TO_TOKENIZER = "/iopsstor/scratch/cscs/asolergi/torchtitan/tokenizer.model"
```
what is this path?

TJ-Solergibert replied:
It's the path to the llama 3 tokenizer. I've exposed all the args through `click`, but I can drop this script if you want!

tianyu-l commented on torchtitan/experiments/multimodal/check_padding_mm.py:
what is this file for? Is it a test, or for sanity check? If so let's put it into a `tests` folder.

TJ-Solergibert replied:
It's just for sanity check. Should I delete it?

tianyu-l commented on torchtitan/experiments/multimodal/llama3_transform.py:
This doesn't seem to be llama3-specific, can we rename to `transform.py` or `clip.py` after we remove `VisionFormatter`.

TJ-Solergibert replied:
Renamed to `transform.py `

tianyu-l reviewed (changes requested):
Thanks for your work. I left some comments.

Please create a new customized `tiktoken.py` within the experiment folder.

TJ-Solergibert commented:
Hi @tianyu-l & @pbontrager, 

The PR is ready for a Re-review! I've addressed all your comments I would say. They've been very helpful, thanks!

In the last push mainly I've created a new `tiktoken.py` under the experiment folder, deleted the `Llama3VisionTransform` class and move all the logic to the dataset itself. I've also refactored a little bit the sanity check script, which basically checks some shapes and the amount of padding:

```
python3 torchtitan/experiments/multimodal/check_padding_mm.py --tokenizer-path tokenizer.model
[titan] 2025-03-31 16:07:48,572 - root - INFO - TikTokenizer built: #words 128257, BOS ID 128000, EOS ID 128001, IMAGE ID 128256
[titan] 2025-03-31 16:07:48,572 - root - INFO - Preparing obelics dataset from HuggingFaceM4/OBELICS
Resolving data files: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████| 1432/1432 [00:00<00:00, 3711.96it/s]
Resolving data files: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████| 1432/1432 [00:00<00:00, 191855.98it/s]

Padding tokens in each sample: tensor([ 31, 194, 262,   0])
Unpadded tokens: 1237, Total tokens in batch: 1724
Padded text tokens: 487, 28.25%
################################################################################
Unpadded images: 18, Total images in batch: 44
Padded images: 26, 59.09% (Each image with shape [4, 3, 448, 448])
################################################################################
Unpadded number of tiles: 55, Total number of tiles: 176
Padded tiles: 121, 68.75% (Each with shape [3, 448, 448])
################################################################################
```

I've left some TODO comments in the code, most of them regarding which arguments we should expose to the user or not:
- `torchtitan/experiments/multimodal/__init__.py`: Which `cls` should we include in the train spec?
- `torchtitan/experiments/multimodal/mm_dataset.py`: Expose `tile_size`, `max_num_tiles`, `image_token`, `image_mean`, `image_std`, `pad_max_tiles` & `padding_idx` through `JobConfig`?
- `torchtitan/experiments/multimodal/tokenizer/tiktoken.py`: Hardcode `IMAGE_TOKEN_ID ` & `IGNORE_INDEX`?

Toni

tianyu-l reviewed (approved):
Looks beautiful! Thank you for building the foundation of multimodal training in torchtitan!

Please fix linting before we could merge.

TJ-Solergibert commented:
Done! Thanks for your comments! Now that I have a bit of bandwidth I will check if I can continue contributing forward! I'm excited to see how far torchtitan gets for multimodal training!
